### PR TITLE
Explicitly define enabled flake8 errors and warnings

### PR DESCRIPTION
**Pull request by ShadowJonathan**
Closes https://github.com/matrix-org/synapse/issues/9366 either way or the other.

I think #9366 has existed for quite some time now, the last action being to enable `B90`, a set of opinionated lints.

However, this was blocked by having all flake8 lints be explicitly defined (using `select=`), as there is no way to "add" a lint code to check. `B90` being optional required this.

This PR could unblock that effort, but I'd want to petition this anyway, as I think now it'd be a good time to look at this again with fresh eyes.

If this PR goes through, I might make an issue about adding `B90`, discussing the pros and cons of those lints.

---

The advantages of doing this would be;
- Having a clear set of lints (as flake8 defaults are hard to find[1]), which could help resolve "is this enabled or not?" questions when adding flake8 plugins in the future.
- Being able to add optional lints (such as `B90`) easily.

Disadvantages would be;
- Flake8 can change its default set of lints inbetween versions, and this would override it.
  - Note that this would only apply if they'd be adding a whole new category, currently it just selects `E`, `W`, and `F`, which are all already the categories flake8 operate in. Any new lints would most likely be added to those, and automatically included here.

[1]: Currently the only way to find flake8 defaults are to run `flake8` with verbose on, and watching a line where it tells about overriding a set of defaults. Here I copied that list as it was per flake8 `4.0.1`.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))
  
`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -9,3 +9,10 @@
 #  E731: do not assign a lambda expression, use a def
 #  E501: Line too long (black enforces this for us)
 ignore=W503,W504,E203,E731,E501
+
+# Default + Enabled flake8 error codes:
```
Comment (richvdh) [inline .flake8:13]: I don't really understand what this sentence is trying to say. Can you rephrase?

File: changelog.d/12600.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Explicitly select the `flake8` lints used.
```
Comment (richvdh) [inline changelog.d/12600.misc:1]: I think this needs clarification too, but currently I'm not quite clear what the actual change is so can't really suggest an improvement.

**Review by richvdh — CHANGES_REQUESTED** [review]

**Review by richvdh — CHANGES_REQUESTED** [review]
I still don't really understand what this is actually changing.

**Comment by ShadowJonathan** [comment]
This is not changing any config to a different behaviour, but it is taking existing implicit configuration, and turning it into something explicit.

It's needed to add something like Bugbear's B90 codes, see [here](https://pypi.org/project/flake8-bugbear/) under "opinionated warnings"

Furthermore, it is also helpful when asking questions like "which flake8 lints are actually enabled?", as the flake8 site does not elaborate on this, and someone would have to;
- set this key in this config to a value (any value)
- run flake8 with verbose on
- hunt a log-line where it says it is overwriting a default set of lints, and grab the ones from there

---

This is not changing anything, but improving on two fronts;
- Allowing flake8 opt-in rules to be easily added
  - This step (of explicitly defining default lines) would have to be done regardless if we wanna add any opt-in lints in the future
- Reduce debugging steps of "why isn't this lint enabled/running?"

**Comment by DMRobertson** [comment]
> the last action being to enable `B90`, a set of opinionated lints.
> 
> However, this was blocked by having all flake8 lints be explicitly defined (using `select=`), as there is no way to "add" a lint code to check. `B90` being optional required this.

Does [`--extend-select`](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-extend-select) not suffice for this?

**Review by DMRobertson — COMMENTED** [review]
I don't see why this is useful.

- We're not spending our time updating the config often, writing new lints for flake8 ourselves; we trust flake8 to provide sensible defaults.
- As of #11537, a fixed version of flake8 is pinned for use in CI, so we are running the same set of defaults and hence the same set of lints everywhere.
- Turning on any additional checks (e.g. the B90x series of lints you mention) should be perfectly possible using the `extend-select` config option as mentioned [here](https://github.com/matrix-org/synapse/pull/12600#issuecomment-1131808123).

Additionally, AFAICS the proposed change is identical to #9370, which we closed last year. I am minded to close this too.

**Comment by ShadowJonathan** [comment]
> Does --extend-select not suffice for this?

...it does, it seems to have been added very recently. This kinda defeats the point of the PR.

I'll close this, in the future I might add B90 lints under extend-select.
